src/cli/commands/generate-types.ts

```diff
@@ -0,0 +1,36 @@
+import { execSync } from "child_process"
+import { program } from "commander"
+import { convertFromDirectory } from "joi-to-typescript"
+
+import { runMain } from "src/cli/utils"
+
+program
+  .description("Generate types from Joi schemas")
+  .requiredOption(
+    "--schemas <schema-dir>",
+    "path to directory containing Joi schemas",
+  )
+  .requiredOption("--out <out-dir>", "path to generated types directory")
+  .option("--no-prettier", "apply prettier to generated files", true)
+  .action((options: { schemas: string; out: string; prettier: boolean }) => {
+    return runMain(async (ctx) => {
+      ctx.logger.info("Generating types")
+
+      await convertFromDirectory({
+        schemaDirectory: options.schemas,
+        indexAllToRoot: true,
+        schemaFileSuffix: "Schema",
+        typeOutputDirectory: options.out,
+        debug: true,
+      })
+
+      if (options.prettier) {
+        ctx.logger.info("Applying prettier to generated files")
+
+        execSync(`prettier --write "${options.out}"`)
```

Comment:
While I don't see quoting becoming a problem here, `spawnSync` (https://nodejs.org/api/child_process.html#child_processspawnsynccommand-args-options) can be used instead as that accepts `string[]` for the arguments, so that way it's not necessary to deal with quoting at all.

Reply:
There's an important difference in `spawn` vs. `exec` behaviour. `exec` throws an error on non-zero exit, which I think is the correct behaviour in this case: 

```
Welcome to Node.js v16.13.1.
Type ".help" for more information.
> child_process.execSync("exit 1")
Uncaught Error: Command failed: exit 1
    at checkExecSyncError (node:child_process:826:11)
    at Object.execSync (node:child_process:900:15) {
  status: 1,
  signal: null,
  output: [ null, <Buffer >, <Buffer > ],
  pid: 70101,
  stdout: <Buffer >,
  stderr: <Buffer >
}
> child_process.spawnSync("exit", ["1"], {shell: true})
{
  status: 1,
  signal: null,
  output: [ null, <Buffer >, <Buffer > ],
  pid: 70228,
  stdout: <Buffer >,
  stderr: <Buffer >
}
```

Reply:
> exec throws an error on non-zero exit, which I think is the correct behaviour in this case

Yes, definitely.

Without the `{shell:true}` option, the result of `spawnSync` contains an `.error`

```console
> node -e 'console.log(require("child_process").spawnSync("exit", ["1"]).error)'
Error: spawnSync exit ENOENT
    at Object.spawnSync (node:internal/child_process:1111:20)
    at Object.spawnSync (node:child_process:814:24)
    at [eval]:1:38
    at Script.runInThisContext (node:vm:129:12)
    at Object.runInThisContext (node:vm:305:38)
    at node:internal/process/execution:76:19
    at [eval]-wrapper:6:22
    at evalScript (node:internal/process/execution:75:60)
    at node:internal/main/eval_string:27:3 {
  errno: -2,
  code: 'ENOENT',
  syscall: 'spawnSync exit',
  path: 'exit',
  spawnargs: [ '1' ]
}
```

So it could be

```js
const result = spawnSync(...)
if (result.error) {
  throw result.error
}
```

I couldn't quite intuit why the `shell: true` option would be needed. Is it for the prettier configuration for that given project to be picked up? If so, the `cwd` option of `spawnSync` would be useful for this.

While it's hard to see how `execSync` could go wrong, `spawnSync` is overall safer in case of malformed input - which I'm not expecting since I imagine this script would be called from a preset command, so either option seems fine.

Reply:
> I know, but execSync is the one call that does it all in this case. You also stated that there could not be any problems with quoting, why do we need to make things more complicated than they are?

We're not _foreseeing_ any problems from how we _expect_ the command to be used. It doesn't mean it's a more reliable option. As I mentioned it leaves the door open for malformed input to have potentially damaging effects and it's not as safe since input isn't being sanitized for the shell.

> binary found, as without shell, PATH resolution is not performed.
> So, I'm using shell here for the same purpose, to resolve prettier binary from PATH.

`spawnSync` also resolves from $PATH. A quick test shows that

```console
$ mkdir tmp
$ cd tmp
$ touch foo
$ chmod +x foo
$ # note PATH="$PWD:$PATH"
$ PATH="$PWD:$PATH" node -e 'console.log(require("child_process").spawnSync("foo"))'
{
  status: 0,
  signal: null,
  output: [ null, <Buffer >, <Buffer > ],
  pid: 681459,
  stdout: <Buffer >,
  stderr: <Buffer >
}
> node -e 'console.log(require("child_process").spawnSync("foo"))'
{
  error: Error: spawnSync foo ENOENT
      at Object.spawnSync (node:internal/child_process:1111:20)
      at Object.spawnSync (node:child_process:814:24)
      at [eval]:1:38
      at Script.runInThisContext (node:vm:129:12)
      at Object.runInThisContext (node:vm:305:38)
      at node:internal/process/execution:76:19
      at [eval]-wrapper:6:22
      at evalScript (node:internal/process/execution:75:60)
      at node:internal/main/eval_string:27:3 {
    errno: -2,
    code: 'ENOENT',
    syscall: 'spawnSync foo',
    path: 'foo',
    spawnargs: []
  },
  status: null,
  signal: null,
  output: null,
  pid: 681721,
  stdout: null,
  stderr: null
}
```

In any case I'm not going to insist on convicing in terms of what's more reliable since I haven't had success with that so far. If you want to keep it like this, ok.

Reply:
Wait, I was wrong. All three of spawn, exec, execFile do resolve PATH

My example didn't work as `exit` is a built-in shell command.  
Unlike `true`, which can be both an executable and built-in command, there's no `/bin/exit`.  
This being said, I'll change it to `execFileSync("prettier", ["--write", options.out])`